base/strings/string.jl

```diff
@@ -106,111 +212,185 @@ function first_utf8_byte(ch::Char)
+        offs += str.len
+    end
+    return out
+end
+
+# UTF-8 encoding length of a character
+function codelen(d::Char)
+    c = UInt32(d)
+    if c < 0x80
+        return 1
+    elseif c < 0x800
+        return 2
+    elseif c < 0x10000
+        return 3
+    elseif c < 0x110000
+        return 4
+    end
+    return 3  # '\ufffd'
 end
 
 function string(a::Union{String,Char}...)
-    s = Array{UInt8}(0)
+    n = 0
+    for d in a
+        if isa(d,Char)
+            n += codelen(d::Char)
+        else
+            n += (d::String).len
+        end
+    end
+    out = _string_n(n)
+    offs = 1
+    p = pointer(out)
     for d in a
         if isa(d,Char)
             c = UInt32(d::Char)
             if c < 0x80
-                push!(s, c%UInt8)
+                unsafe_store!(p, c%UInt8, offs); offs += 1
             elseif c < 0x800
-                push!(s, (( c >> 6          ) | 0xC0)%UInt8)
-                push!(s, (( c        & 0x3F ) | 0x80)%UInt8)
+                unsafe_store!(p, (( c >> 6          ) | 0xC0)%UInt8, offs); offs += 1
+                unsafe_store!(p, (( c        & 0x3F ) | 0x80)%UInt8, offs); offs += 1
             elseif c < 0x10000
-                push!(s, (( c >> 12         ) | 0xE0)%UInt8)
-                push!(s, (((c >> 6)  & 0x3F ) | 0x80)%UInt8)
-                push!(s, (( c        & 0x3F ) | 0x80)%UInt8)
+                unsafe_store!(p, (( c >> 12         ) | 0xE0)%UInt8, offs); offs += 1
+                unsafe_store!(p, (((c >> 6)  & 0x3F ) | 0x80)%UInt8, offs); offs += 1
+                unsafe_store!(p, (( c        & 0x3F ) | 0x80)%UInt8, offs); offs += 1
             elseif c < 0x110000
-                push!(s, (( c >> 18         ) | 0xF0)%UInt8)
-                push!(s, (((c >> 12) & 0x3F ) | 0x80)%UInt8)
-                push!(s, (((c >> 6)  & 0x3F ) | 0x80)%UInt8)
-                push!(s, (( c        & 0x3F ) | 0x80)%UInt8)
+                unsafe_store!(p, (( c >> 18         ) | 0xF0)%UInt8, offs); offs += 1
+                unsafe_store!(p, (((c >> 12) & 0x3F ) | 0x80)%UInt8, offs); offs += 1
+                unsafe_store!(p, (((c >> 6)  & 0x3F ) | 0x80)%UInt8, offs); offs += 1
+                unsafe_store!(p, (( c        & 0x3F ) | 0x80)%UInt8, offs); offs += 1
             else
                 # '\ufffd'
-                push!(s, 0xef); push!(s, 0xbf); push!(s, 0xbd)
+                unsafe_store!(p, 0xef, offs); offs += 1
+                unsafe_store!(p, 0xbf, offs); offs += 1
+                unsafe_store!(p, 0xbd, offs); offs += 1
             end
         else
-            append!(s,(d::String).data)
+            l = (d::String).len
+            unsafe_copy!(pointer(out,offs), pointer(d::String), l)
+            offs += l
         end
     end
-    String(s)
+    return out
 end
 
 function reverse(s::String)
-    dat = s.data
+    dat = convert(Vector{UInt8},s)
     n = length(dat)
     n <= 1 && return s
     buf = Vector{UInt8}(n)
```

Comment:
I would use `_string_n` when passing a pointer is sufficient, and `Vector` when you want vector operations like setindex and `push!`.